### PR TITLE
Code Challenge of bitfriend

### PR DESCRIPTION
**Pull request by bitfriend**
Here is both api project and frontend project.

And please visit here.
### Online Demo
http://160.119.253.137:9091

### CODE REVIEW

**Comment by mat-sf** [comment]
Hi, thank you for the submission, could you please provide us updated keys, because API returns:
`The security token included in the request is invalid`
when I try to run it with the environment keys provided in the example file.

Have a nice day!

**Comment by bitfriend** [comment]
Thank you for your reply.
As you know, amazon key is related to bank card.
I don't love to share my private key via github, because this is public channel and many people can see it.
My email address is bigmanliose@gmail.com.
If you send me email, I will send back my private key via email.

**Comment by bitfriend** [comment]
### Online Demo
Already deployed this app here.
http://160.119.253.137:9091

File: frontend/src/components/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,307 @@
+import React, { FunctionComponent, MouseEvent, useEffect, useState } from 'react';
+import styled from 'styled-components';
+import { motion, AnimatePresence } from 'framer-motion';
+import { useMutation } from '@apollo/client';
+import { isEmpty } from 'lodash';
```
Comment (mat-sf) [inline frontend/src/components/Modal.tsx:5]: Do u see any problems with importing function like that(tree shaking)?

File: frontend/src/components/Modal.tsx [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+import GMap from './GMap';
+
+interface ModalProps {
+  open: boolean;
+  user: User | null;
+  onClose: () => void;
+}
+
+const Modal: FunctionComponent<ModalProps> = ({ open, user, onClose }) => {
```
Comment (mat-sf) [inline frontend/src/components/Modal.tsx:19]: How would you make that modal accessible?

File: frontend/src/components/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+interface PageButtonProps {
+  foreColor: string;
+  backColor: string;
+  active?: boolean;
+}
+
+const PageButton = styled.button<PageButtonProps>`
+  background: transparent;
+  border: none;
+  height: 2rem;
```
Comment (mat-sf) [inline frontend/src/components/Pagination.tsx:125]: Can you please elaborate a little bit on units you use - pixels for font-size/margin/padding, rems for heigh/width - why? Any strategy here?

File: frontend/src/components/GMap.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  });
+
+  const onClick = (e: google.maps.MapMouseEvent) => {
+    setClicks([...clicks, e.latLng!]);
+  }
+
+  const onIdle = (m: google.maps.Map) => {
+    console.log('onIdle', m);
+    // setZoom(m.getZoom());
+    // setCenter(m.getCenter().toJSON());
```
Comment (mat-sf) [inline frontend/src/components/GMap.tsx:42]: WDYT about having commented code in your projects?

File: frontend/src/components/Pagination.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    box-shadow: transparent 0px 0px 0px 1px, transparent 0px 0px 0px 4px, rgba(0, 0, 0, 0.12) 0px 6px 16px;
+  }
+`;
+
+const Separator = styled.div`
+  width: 1rem;
+  margin: 0 0.25rem;
+`;
+
+export default Pagination;
```
Comment (mat-sf) [inline frontend/src/components/Pagination.tsx:154]: Any thoughts on default vs named exports? Which one do you prefer, any pros/cons on one over the other?

**Review by mat-sf — COMMENTED** [review]
Hi again! I connected with my AWS so no worries.
Thank you for deploying the app and the time you have spent on this.

I left a few comments so we can have a little discussion, if you have any questions - feel free to ask 🙂

File: frontend/src/components/Modal.tsx (hunk 1 above)
Comment (bitfriend) [reply frontend/src/components/Modal.tsx:5]: Yes, you are correct.
Tree-shaking is syntax-sensitive. To enable tree-shaking in lodash, I have to import functions using syntax like `import isEmpty from 'lodash/isEmpty'`. `import { isEmpty } from 'lodash'` will not tree-shake, nor will, obviously, `import _ from 'lodash'`.
Thank you for your advice.

File: frontend/src/components/Pagination.tsx (hunk 3 above)
Comment (bitfriend) [reply frontend/src/components/Pagination.tsx:125]: I thought `media-query` doesn’t apply in this case.
The data shown doesn’t exceed the minimum screen width so I left out pagination.

File: frontend/src/components/Modal.tsx (hunk 2 above)
Comment (bitfriend) [reply frontend/src/components/Modal.tsx:19]: This modal shows up when `open` is `true`.
It is when this pencil icon is clicked.
image

File: frontend/src/components/Pagination.tsx (hunk 5 above)
Comment (bitfriend) [reply frontend/src/components/Pagination.tsx:154]: Named Exports
    Can export multiple values.
    Must use the exported name when importing. If you want to change name, use `as` modifier.

Default Exports
    Export a single value.
    Can use any name when importing.

File: frontend/src/components/GMap.tsx (hunk 4 above)
Comment (bitfriend) [reply frontend/src/components/GMap.tsx]: Just deleted the unnecessary code and pushed to github.
Please review new commits.

**Comment by bitfriend** [comment]
> Hi again! I connected with my AWS so no worries. Thank you for deploying the app and the time you have spent on this.
> 
> I left a few comments so we can have a little discussion, if you have any questions - feel free to ask 🙂

**Comment by bitfriend** [comment]
I closed my pull request by mistake.
So I reopened it.

**Pull request closed without merging**